### PR TITLE
Added changes for authentication and logout flows to handle SAML single logout requests from federated IdPs

### PR DESCRIPTION
**Pull request by Isanka95**
### Proposed changes in this pull request

- Resolves: wso2/product-is#5717

- Added DB operations into UserSessionStore class to map the session context key with the SAML federated idp session index and delete those details after the logout. 

- Used the authenticationStepHistory attribute in the authentication context object to store the federated idp session index.     

- authenticationStepHistory attribute includes the list of AuthHistory objects and added the idp session index as an attribute of it.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/context/AuthHistory.java [hunk 1]
```diff
@@ -47,6 +48,12 @@ public AuthHistory(String authenticatorName, String idpName, String code) {
         this.code = code;
     }
 
+    public AuthHistory(String authenticatorName, String idpName, Object idpSessionIndexObj) {
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/context/AuthHistory.java]: add new line here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 2]
```diff
@@ -471,6 +474,22 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
                             "the database", e);
                 }
             }
+            // Check whether the authentication flow includes a SAML federated IdP and then
+            // store the federated idp index with the session context key for the single logout.
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: store the federated idp index --> saml session index

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 3]
```diff
@@ -471,6 +474,22 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
...
+            if (context.getAuthenticationStepHistory() != null) {
+                for (AuthHistory authHistory : context.getAuthenticationStepHistory()) {
+                    if ((FED_AUTH_NAME).equals(authHistory.getAuthenticatorName()) &&
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: remove unnecessary brackets added here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java [hunk 4]
```diff
@@ -98,6 +105,22 @@ public void handle(HttpServletRequest request, HttpServletResponse response, Aut
             }
         }
 
+        // Remove federated authentication session details from the database.
+        if ((sessionContext != null) && StringUtils.isNotBlank(context.getSessionIdentifier()) &&
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java]: remove unnecessary brackets

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java [hunk 5]
```diff
@@ -98,6 +105,22 @@ public void handle(HttpServletRequest request, HttpServletResponse response, Aut
...
+                (sessionContext.getSessionAuthHistory().getHistory() != null)) {
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java]: remove unnecessary brackets

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java [hunk 6]
```diff
@@ -98,6 +105,22 @@ public void handle(HttpServletRequest request, HttpServletResponse response, Aut
...
+            for (AuthHistory authHistory : sessionContext.getSessionAuthHistory().getHistory()) {
+                if (FED_AUTH_NAME.equals(authHistory.getAuthenticatorName())) {
+                    try {
+                        UserSessionStore.getInstance().removeFederatedAuthSessionInfo(context.getSessionIdentifier());
+                        break;
+                    } catch (UserSessionException e) {
+                        throw new FrameworkException("Error while deleting federated authentication session details for"
+                                + " the session contect key :" + context.getSessionIdentifier(), e);
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java]: contect --> context

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 7]
```diff
@@ -555,7 +559,22 @@ protected void doAuthentication(HttpServletRequest request, HttpServletResponse
...
+            // Add SAML federated idp session index into the authentication step history.
+            String idpSessionIndex = null;
+            String parameterName = FEDERATED_IDP_SESSION_ID + idpName;
+            AuthHistory authHistory = new AuthHistory(authenticator.getName(), idpName);
+
+            if (context.getParameters().containsKey(parameterName)) {
+                idpSessionIndex = context.getParameter(parameterName).toString();
+            }
+            if (StringUtils.isNotBlank(context.getCurrentAuthenticator()) && StringUtils.isNotBlank(idpSessionIndex)) {
+                if (FED_AUTH_NAME.equals(context.getCurrentAuthenticator())) {
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: improve this logic by combining these two if cases

Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Improve this logic by combining these two if cases.

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql [hunk 8]
```diff
@@ -1422,6 +1422,17 @@ CREATE TABLE IDN_FUNCTION_LIBRARY (
 )
 /
 
+CREATE TABLE IDN_FEDERATED_AUTH_SESSION_MAPPING (
+	IDP_SESSION_ID VARCHAR(255) NOT NULL,
+	SESSION_ID VARCHAR(255) NOT NULL,
```
Comment (thanujalk) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql]: create an index for the SESSION_ID.

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql [hunk 9]
```diff
@@ -1422,6 +1422,17 @@ CREATE TABLE IDN_FUNCTION_LIBRARY (
...
+	IDP_NAME VARCHAR(255) NOT NULL,
+	AUTHENTICATOR_ID VARCHAR(255),
+	PROTOCOL_TYPE VARCHAR(255),
```
Comment (thanujalk) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql]: Can these two be null?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java [hunk 10]
```diff
@@ -81,11 +88,11 @@ public void handle(HttpServletRequest request, HttpServletResponse response, Aut
...
         }
         SequenceConfig sequenceConfig = context.getSequenceConfig();
+        // Retrieve session information from cache.
+        SessionContext sessionContext = FrameworkUtils.getSessionContextFromCache(context.getSessionIdentifier());
         ExternalIdPConfig externalIdPConfig = null;
         if (FrameworkServiceDataHolder.getInstance().getAuthnDataPublisherProxy() != null &&
                 FrameworkServiceDataHolder.getInstance().getAuthnDataPublisherProxy().isEnabled(context)) {
-            // Retrieve session information from cache in order to publish event
-            SessionContext sessionContext = FrameworkUtils.getSessionContextFromCache(context.getSessionIdentifier());
             if (sessionContext != null) {
```
Comment (thanujalk) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java]: Can merge these if conditions

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 11]
```diff
@@ -685,4 +686,46 @@ public boolean isExistingMapping(User user, int idpId, String sessionId) throws
...
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+             PreparedStatement prepStmt = connection.prepareStatement(SQLQueries.SQL_STORE_FEDERATED_AUTH_SESSION_INFO)) {
+            prepStmt.setString(1, authHistory.getIdpSessionIndex());
+            prepStmt.setString(2, sessionContextKey);
+            prepStmt.setString(3, authHistory.getIdpName());
+            prepStmt.setString(4, authHistory.getAuthenticatorName());
+            prepStmt.setString(5, authHistory.getRequestType());
+            prepStmt.execute();
+        } catch (SQLException e) {
+            throw new UserSessionException("Error while adding session data for key:" + sessionContextKey, e);
```
Comment (thanujalk) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java]: add Idp name as well... in the log.

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql (hunk 9 above)
Comment (Isanka95) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql]: I'm not using the values of these two fields. Added these two for future work.

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql [hunk 12]
```diff
@@ -1530,3 +1541,7 @@ CREATE INDEX IDX_AI_DN_UN_AI ON IDN_ASSOCIATED_ID(DOMAIN_NAME, USER_NAME, ASSOCI
 -- IDN_OAUTH2_TOKEN_BINDING --
 CREATE INDEX IDX_IDN_AUTH_BIND ON IDN_OAUTH2_TOKEN_BINDING (TOKEN_BINDING_REF)
   /
+
+-- IDN_FEDERATED_AUTH_SESSION_MAPPING --
+CREATE INDEX IDX_FEDERATED_AUTH_SESSION_ID ON IDN_FEDERATED_AUTH_SESSION_MAPPING (SESSION_ID)
+  /
```
Comment (thanujalk) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/db2.sql]: add a new line here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java [hunk 13]
```diff
@@ -81,20 +88,35 @@ public void handle(HttpServletRequest request, HttpServletResponse response, Aut
...
-            // Retrieve session information from cache in order to publish event
-            SessionContext sessionContext = FrameworkUtils.getSessionContextFromCache(context.getSessionIdentifier());
-            if (sessionContext != null) {
-                Object authenticatedUserObj = sessionContext.getProperty(FrameworkConstants.AUTHENTICATED_USER);
-                AuthenticatedUser authenticatedUser = new AuthenticatedUser();
-                if (authenticatedUserObj != null) {
-                    authenticatedUser = (AuthenticatedUser) authenticatedUserObj;
+                FrameworkServiceDataHolder.getInstance().getAuthnDataPublisherProxy().isEnabled(context) &&
+                    sessionContext != null) {
+            Object authenticatedUserObj = sessionContext.getProperty(FrameworkConstants.AUTHENTICATED_USER);
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java:97]: Why can't we directly use `AuthenticatedUser authenticatedUserObj`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 14]
```diff
@@ -555,7 +559,21 @@ protected void doAuthentication(HttpServletRequest request, HttpServletResponse
...
             //add authenticated idp data to the session wise map
             context.getCurrentAuthenticatedIdPs().put(idpName, authenticatedIdPData);
-            context.addAuthenticationStepHistory(new AuthHistory(authenticator.getName(), idpName));
+
+            // Add SAML federated idp session index into the authentication step history.
+            String idpSessionIndex = null;
+            String parameterName = FEDERATED_IDP_SESSION_ID + idpName;
+            AuthHistory authHistory = new AuthHistory(authenticator.getName(), idpName);
+
+            if (context.getParameters().containsKey(parameterName)) {
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:568]: Can't `context.getParameters()` e null? if so do the null check

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 15]
```diff
@@ -685,4 +686,47 @@ public boolean isExistingMapping(User user, int idpId, String sessionId) throws
...
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+             PreparedStatement prepStmt = connection.prepareStatement(SQLQueries.SQL_STORE_FEDERATED_AUTH_SESSION_INFO)) {
+            prepStmt.setString(1, authHistory.getIdpSessionIndex());
+            prepStmt.setString(2, sessionContextKey);
+            prepStmt.setString(3, authHistory.getIdpName());
+            prepStmt.setString(4, authHistory.getAuthenticatorName());
+            prepStmt.setString(5, authHistory.getRequestType());
+            prepStmt.execute();
+        } catch (SQLException e) {
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:709]: Don't we need a rollback here?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 16]
```diff
@@ -685,4 +686,47 @@ public boolean isExistingMapping(User user, int idpId, String sessionId) throws
...
+     * @param sessionContextKey Session Context Key.
+     * @throws UserSessionException Error while deleting session details of a given session id.
+     */
+    public void removeFederatedAuthSessionInfo(String sessionContextKey) throws UserSessionException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+             PreparedStatement prepStmt = connection.prepareStatement(SQLQueries.SQL_DELETE_FEDERATED_AUTH_SESSION_INFO)) {
+            prepStmt.setString(1, sessionContextKey);
+            prepStmt.execute();
+        } catch (SQLException e) {
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:727]: Don't we need a rollback here?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java [hunk 17]
```diff
@@ -81,20 +88,35 @@ public void handle(HttpServletRequest request, HttpServletResponse response, Aut
...
+                authenticatedUser = (AuthenticatedUser) authenticatedUserObj;
+            }
+            FrameworkUtils.publishSessionEvent(context.getSessionIdentifier(), request, context,
+                    sessionContext, authenticatedUser, FrameworkConstants.AnalyticsAttributes
+                            .SESSION_TERMINATE);
+        }
+
+        // Remove federated authentication session details from the database.
+        if (sessionContext != null && StringUtils.isNotBlank(context.getSessionIdentifier()) &&
+                sessionContext.getSessionAuthHistory().getHistory() != null) {
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java:109]: Can't the `sessionContext.getSessionAuthHistory()` e null? If so do the null check?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 18]
```diff
@@ -471,6 +474,22 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
                             "the database", e);
                 }
             }
+            // Check whether the authentication flow includes a SAML federated IdP and
+            // store the saml index with the session context key for the single logout.
+            if (context.getAuthenticationStepHistory() != null) {
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:479]: What will happen if context.getAuthenticationStepHistory() is null? can we add some debug logs for that

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 19]
```diff
@@ -471,6 +474,22 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
...
+                for (AuthHistory authHistory : context.getAuthenticationStepHistory()) {
+                    if (FED_AUTH_NAME.equals(authHistory.getAuthenticatorName()) &&
+                            StringUtils.isNotBlank(authHistory.getIdpSessionIndex()) &&
+                            StringUtils.isNotBlank(authHistory.getIdpName())) {
+                        try {
+                            UserSessionStore.getInstance().storeFederatedAuthSessionInfo(sessionContextKey, authHistory);
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:485]: can we add some debug logs for storing process with some context, which IdP is being stored and what sessionIdentifier

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java (hunk 13 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java:97]: https://github.com/wso2/carbon-identity-framework/pull/3738

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 14 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:568]: Fixed with PR https://github.com/wso2/carbon-identity-framework/pull/3738

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java (hunk 16 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:727]: Fixed with PR https://github.com/wso2/carbon-identity-framework/pull/3738

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java (hunk 15 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:709]: Fixed with PR https://github.com/wso2/carbon-identity-framework/pull/3738

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java (hunk 17 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java:109]: Fixed with PR https://github.com/wso2/carbon-identity-framework/pull/3738

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java (hunk 18 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:479]: Fixed with PR https://github.com/wso2/carbon-identity-framework/pull/3738

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java (hunk 19 above)
Comment (IsankaSR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:485]: Fixed with PR https://github.com/wso2/carbon-identity-framework/pull/3738

**Pull request merged**